### PR TITLE
Reward weights + bug fixes

### PR DESCRIPTION
**Pull request by p-ferreira**
This PR adds the functionality of setting `model_weights` to the reward models accordingly to defined values in `openvalidators/reward/config.py` or flags like `--reward.diversity_weights`. 

Also includes:

- adds reward tensor to `self.` to be used in `forward` pass reward calculation
- modifies boolean tags such `--neuron.reciprocate_off` to weight flags like `--reward.reciprocate_weights`
- adds `RewardModelType` enum for key normalization and organization
- adds `DefaultRewardFrameworkConfig` class to define default reward weights
- sets Mock reward models to return zeros instead of ones
- fix `MockDataset` naming mismatch bug introduced in latest dataset change

### CODE REVIEW

File: openvalidators/reward/reward.py [hunk 1]
```diff
@@ -104,7 +105,7 @@ def apply( self, prompt: str, responses: List[ bt.DendriteCall ], name: str) ->
 
         # Fill reward tensor.
         for idx, reward in zip(successful_completions_indices, successful_rewards):
-            filled_rewards[idx] = reward
+            filled_rewards[idx] = reward * self.model_weight
```
Comment (opentaco) [inline openvalidators/reward/reward.py:108]: It may be more fitting to move the weighting into the weighted sum in forward.py:
```python
for weight_i, reward_fn_i in zip(self.reward_weights, self.reward_functions):
        reward_i = reward_fn_i.apply( prompt, responses, name ).to( self.device )
        rewards += weight_i * reward_i
```
Check this: https://github.com/opentensor/validators/compare/sequence_rewards...feature/reward-stack-weights?expand=1

File: openvalidators/reward_config.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# The 'reward_models_weights' configuration sets the distribution of weights used in the reward framework.
+# These weights determine the relative importance of each reward model.
+# It is important to note that all the weights must sum up to one (1) to ensure a proper balance in the reward calculations.
+#
+# Note: Fields should match RewardModelType Enum in openvalidators/config.py
+reward_models_weights:
```
Comment (opentaco) [inline openvalidators/reward_config.yaml:6]: rlhf supercedes dahoas (so turn this off), and also outperforms prompt-based (so turn this off).
So the following weights may be safer for now:
```python
reward_models_weights:
   rlhf_reward_model: 0.4
   reciprocate_reward_model: 0.3
   dahoas_reward_model: 0
   diversity_reward_model: 0.3
   prompt_reward_model: 0
```

File: openvalidators/neuron.py [hunk 3]
```diff
@@ -131,22 +133,31 @@ def __init__(self):
...
-                OpenAssistantRewardModel( device = self.device ) if not self.config.neuron.openassistant_off else MockRewardModel('rlhf_reward_model'), 
-                ReciprocateRewardModel( device = self.device ) if not self.config.neuron.reciprocate_off else MockRewardModel('reciprocate_reward_model'),
-                DahoasRewardModel( path = self.config.neuron.full_path, device = self.device ) if not self.config.neuron.dahoas_off else MockRewardModel('dahoas_reward_model'),
-                DiversityRewardModel( device = self.device ) if not self.config.neuron.diversity_off else MockRewardModel('diversity_reward_model'),
-                PromptRewardModel( device = self.device ) if not self.config.neuron.prompt_based_off else MockRewardModel('prompt_reward_model'),
+            reward_config = RewardFrameworkConfig(config_file_path=self.config.neuron.reward_config_path)
+
+            self.reward_functions = [
+                OpenAssistantRewardModel(device=self.device, model_weight=reward_config.rlhf_model_weight)
+                    if not self.config.neuron.openassistant_off else MockRewardModel(RewardModelType.rlhf.value),
```
Comment (opentaco) [inline openvalidators/neuron.py:145]: A potential problem with having weight separate from the off flag is that if the weight is 0 then the model should ideally be off, despite what the flag specifies.
It may be more natural to just specify a weight config and determine if the model should be off/on.
See: https://github.com/opentensor/validators/compare/sequence_rewards...feature/reward-stack-weights?expand=1

File: openvalidators/reward/reward.py (hunk 1 above)
Comment (p-ferreira) [reply openvalidators/reward/reward.py:108]: I think the difference in the approaches is where the responsibility of the model weights should be. I took the approach of having the model as responsible for its own weighted sum (weight in), whereas what you propose is the “framework” (self) to bear this responsibility and explicitly define that in the sum (weight out).

I think both approaches are valid but I like your approach, will change the code accordingly

File: openvalidators/neuron.py (hunk 3 above)
Comment (p-ferreira) [reply openvalidators/neuron.py:145]: The implicit rule of `--model-flag=0` == **do not start the model** seems less natural to me than `--model-flag=on/off` where the configs are isolated into a centralized point of focus.

The rationale behind this implementation is that having multiple approaches (file and flags) adds to the point of failures and increases complexity, so in the choice on one of them, I thought that a config file would be more transparent than scattered flags throughout the code.

But considering your other comments I think we can take a different route on this one, so following your idea, will push some other design

**Review by opentaco — APPROVED** [review]
With the latest changes we have the best of both approaches, solid work, thanks.

File: openvalidators/reward/open_assistant.py [hunk 4]
```diff
@@ -25,7 +26,7 @@ class OpenAssistantRewardModel( BaseRewardModel ):
     reward_model_name: str = "OpenAssistant/reward-model-deberta-v3-large-v2"
 
     @property
-    def name(self) -> str: return "rlhf_reward_model"
+    def name(self) -> str: return RewardModelType.rlhf.value
```
Comment (Eugene-hu) [inline openvalidators/reward/open_assistant.py:29]: Can we rename this model to openassistant?

File: openvalidators/reward/config.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    nsfw = 'nsfw_filter'
+    relevance = 'relevance_filter'
+
+
+@dataclass(frozen=True)
+class DefaultRewardFrameworkConfig:
+    """Reward framework default configuration.
+    Note: All the weights should add up to 1.0.
+    """
+    rlhf_model_weight: float = 0.4
```
Comment (Eugene-hu) [inline openvalidators/reward/config.py:38]: I think by default lets only turn on a single reward model. In this case, once we are sure that all of the reward models work in combination, then we can change the default behaviour

Comment (Eugene-hu) [reply openvalidators/reward/config.py:38]: My recommendation right now is to only turn on the openassistant model, but that may change in the coming days

**Review by Eugene-hu — APPROVED** [review]
LGTM, left you a few comments

File: openvalidators/reward/open_assistant.py (hunk 4 above)
Comment (p-ferreira) [reply openvalidators/reward/open_assistant.py:29]: Don't you believe RLHF represents more transparently the type of reward model that is being used?
That way we abstract at the model type level and reduce time of implementation change if we ever want to use other rlhf based model.

Since things move very fast I thought it would be better to be decoupled from the model name and being coupled to framework functionality instead. What do think?

File: openvalidators/reward/config.py (hunk 5 above)
Comment (p-ferreira) [reply openvalidators/reward/config.py:38]: Got it, will update this to set 1.0 on the openassistant model and 0 to the rest

File: openvalidators/reward/config.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    nsfw = 'nsfw_filter'
+    relevance = 'relevance_filter'
+
+
+@dataclass(frozen=True)
+class DefaultRewardFrameworkConfig:
+    """Reward framework default configuration.
+    Note: All the weights should add up to 1.0.
+    """
+    rlhf_model_weight: float = 1.0
```
Comment (steffencruz) [inline openvalidators/reward/config.py:38]: Does it make sense to include `nsfw` and `blacklist` here for completeness?

File: openvalidators/event.py [hunk 7]
```diff
@@ -32,14 +51,14 @@ class EventSchema:
...
         rewards = {
-            'dahoas_reward_model': event_dict.get('dahoas_reward_model'),
-            'blacklist_filter': event_dict.get('blacklist_filter'),
-            'nsfw_filter': event_dict.get('nsfw_filter'),
-            'reciprocate_reward_model': event_dict.get('reciprocate_reward_model'),
-            'diversity_reward_model': event_dict.get('diversity_reward_model'),
-            'rlhf_reward_model': event_dict.get('rlhf_reward_model'),
-            'prompt_reward_model': event_dict.get('prompt_reward_model'),
-            'relevance_scoring': event_dict.get('relevance_scoring'),
+            'dahoas_reward_model': event_dict.get(RewardModelType.dahoas.value),
```
Comment (steffencruz) [inline openvalidators/event.py:54]: IMO keys should be coming from the `RewardModelType` class too. This will ensure that naming conventions are kept consistent.

```python
In [28]: event_dict = {'rlhf_reward_model':[1,2,3], 'dahoas_reward_model': [5,3,1], 'nsfw_filter': [69,69,69]}

In [29]: {reward_model.name: event_dict.get(reward_model.value) for reward_model in RewardModelType}
Out[29]: 
{'rlhf': [1, 2, 3],
 'reciprocate': None,
 'dahoas': [5, 3, 1],
 'diversity': None,
 'prompt': None,
 'blacklist': None,
 'nsfw': [69, 69, 69],
 'relevance': None}
```

File: openvalidators/neuron.py [hunk 8]
```diff
@@ -130,23 +130,52 @@ def __init__(self):
         bt.logging.debug("loading", "reward_functions")
         if self.config.neuron.mock_reward_models:
             self.reward_functions = []
+            self.reward_weights = []
             self.masking_functions = [
-                MockRewardModel('blacklist_filter'), 
-                MockRewardModel('nsfw_filter') 
+                MockRewardModel(RewardModelType.blacklist.value),
+                MockRewardModel(RewardModelType.nsfw.value),
```
Comment (steffencruz) [inline openvalidators/neuron.py:136]: Why is `nsfw` a masking function rather than a reward function?

File: openvalidators/event.py (hunk 7 above)
Comment (steffencruz) [reply openvalidators/event.py:54]: BTW I'm unable to comment on the lines beyond those which were edited in this PR but I'm seeing that the reward/filter models in the event class have `Optional` type hints but no default values.

File: openvalidators/reward/open_assistant.py (hunk 4 above)
Comment (p-ferreira) [reply openvalidators/reward/open_assistant.py:29]: Super agree. BTW, @steffencruz @Eugene-hu @opentaco do you know which type of model this dahoas instance is? 
The [model card](https://huggingface.co/Dahoas/gptj-rm-static) on github doesn't tell much

File: openvalidators/neuron.py [hunk 9]
```diff
@@ -130,23 +130,52 @@ def __init__(self):
...
+            # Ensure reward function weights sum to 1.
+            if self.reward_weights.sum() != 1:
+                message = f"Reward function weights do not sum to 1 (Current sum: {self.reward_weights.sum()}.)" \
+                          f"Check your reward config file at `reward/config.py` or ensure that all your cli reward flags sum to 1."
+                bt.logging.error(message)
+                raise Exception(message)
+
+            self.reward_functions = [
+                OpenAssistantRewardModel(device=self.device) if self.config.reward.rlhf_weight > 0
+                    else MockRewardModel(RewardModelType.rlhf.value),
```
Comment (steffencruz) [inline openvalidators/neuron.py:154]: Really like the reference to the `RewardModelType` class here

File: openvalidators/reward/config.py (hunk 6 above)
Comment (p-ferreira) [reply openvalidators/reward/config.py:38]: It was there initially but their are being instantiated and configured in different ways. `nsfw` and `blacklist` are masking models and they do not have weighting due to that

File: openvalidators/event.py (hunk 7 above)
Comment (p-ferreira) [reply openvalidators/event.py:54]: Yeah, I usually go with the approach `Less code = Better code` but for tests I prefer to go with `Explicit code > Implicit code` for clarity on what's going on. On line 120 you can see that I use this very same "Less code" approach , as the rationale of the test is specifically on that. For this particular one I thought 8 LOCs for more clarity woudn't be a big problem

File: openvalidators/reward/open_assistant.py (hunk 4 above)
Comment (steffencruz) [reply openvalidators/reward/open_assistant.py:29]: I don't know. Ha, the model card doesn't help a bit.

File: openvalidators/neuron.py (hunk 8 above)
Comment (p-ferreira) [reply openvalidators/neuron.py:136]: I think @Eugene-hu and @isabella618033 will know to answer this better than me. From what I understood, they are played out like a binary filter, where it either returns 0 or 1, but not sure if that's the case.

File: openvalidators/event.py (hunk 7 above)
Comment (steffencruz) [reply openvalidators/event.py:54]: Current solution is ok. Future me is kicking off about it because we are susceptible to naming convention drift, which means that dashboard and any analytics layer is going to be brittle w.r.t. schema changes. 

It's not a hill I'm prepared to die on. After all it's only monday 😆

File: openvalidators/neuron.py (hunk 8 above)
Comment (steffencruz) [reply openvalidators/neuron.py:136]: We actually have a NSFW threshold parameter `bound` which maps the float value to 0 or 1. Would be good to have the raw values in there for deeper introspection about flagged completions

File: openvalidators/neuron.py [hunk 10]
```diff
@@ -130,23 +130,52 @@ def __init__(self):
...
-                DiversityRewardModel( device = self.device ) if not self.config.neuron.diversity_off else MockRewardModel('diversity_reward_model'),
-                PromptRewardModel( device = self.device ) if not self.config.neuron.prompt_based_off else MockRewardModel('prompt_reward_model'),
+            self.reward_weights = torch.tensor([
+                self.config.reward.rlhf_weight, self.config.reward.reciprocate_weight, self.config.reward.dahoas_weight,
+                self.config.reward.diversity_weight, self.config.reward.prompt_based_weight
+            ], dtype=torch.float32).to(self.device)
+
+            # Ensure reward function weights sum to 1.
+            if self.reward_weights.sum() != 1:
+                message = f"Reward function weights do not sum to 1 (Current sum: {self.reward_weights.sum()}.)" \
```
[the remark this answers is not in the document]
Comment (p-ferreira) [reply openvalidators/neuron.py:147]: Yeah, that was part of the conversation we had last week, but in some moment @Eugene-hu brought the same idea of maybe implementing a softmax on this sum.

Do you believe that this approach would be more natural to a end-user than normalized units?
On my end I believe we cannot go more transparent than that, all the values should be sum to one, no space for interpretation, but if you think the rationale of non-normalized vs normalized of the end validator user will have that big of impact in the definition of the framework, we can discuss alternatives

File: openvalidators/neuron.py (hunk 9 above)
Comment (steffencruz) [reply openvalidators/neuron.py:154]: Gotcha

File: openvalidators/neuron.py (hunk 8 above)
Comment (p-ferreira) [reply openvalidators/neuron.py:136]: I like this idea, I think we can create a new issue to map this feature

File: openvalidators/reward/open_assistant.py (hunk 4 above)
Comment (Eugene-hu) [reply openvalidators/reward/open_assistant.py:29]: Correct me if I am wrong, but aren't both dahoas and relevance trained to act as the reward model in an RLHF system? Just quickly skimming through the model cards, nothing jumps out to me.

Comment (p-ferreira) [reply openvalidators/reward/open_assistant.py:29]: I don't know about the dahoas model due to lack of information, but it seems that the relevance model is just a BERT (bert-base-uncased) model calculating the RMSE distance for the mean polled embeddings of the completion and the prompt.
